pkg/volume/util/operationexecutor/operation_generator.go

```diff
@@ -627,6 +627,262 @@ func (og *operationGenerator) GenerateUnmountDeviceFunc(
+		volumeToMount.VolumeSpec,
+		volumeToMount.Pod,
+		volumeToMount.DevicePath,
+		volume.VolumeOptions{})
+	if newMapperErr != nil {
+		eventErr, detailedErr := volumeToMount.GenerateError("MapVolume.NewBlockVolumeMapper initialization failed", newMapperErr)
+		og.recorder.Eventf(volumeToMount.Pod, v1.EventTypeWarning, kevents.FailedMapVolume, eventErr.Error())
+		return nil, blockVolumePlugin.GetPluginName(), detailedErr
+	}
+
+	// Get attacher, if possible
+	attachableVolumePlugin, _ :=
+		og.volumePluginMgr.FindAttachablePluginBySpec(volumeToMount.VolumeSpec)
+	var volumeAttacher volume.Attacher
+	if attachableVolumePlugin != nil {
+		volumeAttacher, _ = attachableVolumePlugin.NewAttacher()
+	}
+
+	return func() error {
+		if volumeAttacher != nil {
+			// Wait for attachable volumes to finish attaching
+			glog.Infof(volumeToMount.GenerateMsgDetailed("MapVolume.WaitForAttach entering", fmt.Sprintf("DevicePath %q", volumeToMount.DevicePath)))
+
+			devicePath, err := volumeAttacher.WaitForAttach(
+				volumeToMount.VolumeSpec, volumeToMount.DevicePath, volumeToMount.Pod, waitForAttachTimeout)
+			if err != nil {
+				// On failure, return error. Caller will log and retry.
+				return volumeToMount.GenerateErrorDetailed("MapVolume.WaitForAttach failed", err)
+			}
+
+			glog.Infof(volumeToMount.GenerateMsgDetailed("MapVolume.WaitForAttach succeeded", ""))
+
+			// Set up for mapping device to global map path using symbolic link
+			globalVDPDPath, err :=
+				blockVolumeMapper.GetVolumeDeviceMapPath(volumeToMount.VolumeSpec)
+			if err != nil {
+				// On failure, return error. Caller will log and retry.
+				return volumeToMount.GenerateErrorDetailed("MapVolume.GetDeviceMountPath failed", err)
+			}
+
+			// Map device to global map path under the given plugin directory using symbolic link
+			mapErr := util.MapDevice(devicePath, globalVDPDPath, volumeToMount.Pod.UID)
+			if mapErr != nil {
+				// On failure, return error. Caller will log and retry.
+				eventErr, detailedErr := volumeToMount.GenerateError("MapVolume.SetUp failed", mapErr)
+				og.recorder.Eventf(volumeToMount.Pod, v1.EventTypeWarning, kevents.FailedMapVolume, eventErr.Error())
+				return detailedErr
+			}
+			// Device mapping for global map path succeeded
+			glog.Infof(volumeToMount.GenerateMsgDetailed("MapVolume.MapDevice succeeded", fmt.Sprintf("globalVDPDPath %q", globalVDPDPath)))
+
+			// Update actual state of world to reflect volume is globally mounted
+			markDeviceMappedErr := actualStateOfWorld.MarkDeviceAsMounted(
+				volumeToMount.VolumeName)
+			if markDeviceMappedErr != nil {
+				// On failure, return error. Caller will log and retry.
+				return volumeToMount.GenerateErrorDetailed("MapVolume.MarkDeviceAsMounted failed", markDeviceMappedErr)
+			}
+		}
+
+		// Map device to volume map path under the given pod directory using symbolic link
+		// A plugin doesn't have attacher also needs to map device to global map path with SetUpDevice()
```

Comment:
It's confusing that we have different requirements for SetUp depending on the plugin.  We should have consistent requirements for implementing the interface.

Reply:
OK.
If we should have consistent requirement to SetUpDevice(), the method should create symbolic link on both global map path and volume map path regardless of whether the plugin have attacher or not.
Is this make sense? 
Then we will remove MapDevice() and UnmapDevice() from operation_executor.

Reply:
Would it work if we made a symlink from pod's volume directory to the global map, after calling SetUp?  Then all plugins only have to take care of setting up the volume at the pod's path.

Reply:
If the SetUpDevice only take care of volume setting up, I suppose GenerateMapVolumeFunc() needs following steps,

1. SetUpDevice() -> this should return block device path
2. Create global map path and symlink to block device.
3. Create pod's volume directory and symlink from pod's volume directory to block device(or global map?)

Reply:
or may be this also work.

1. SetUpDevice() -> only take care of volume setup, return error status.
2. Get block device path from mapper.GetDevicePath() 
3. Create global map path directory and symlink to block device.
4. Create pod's volume directory and symlink from pod's volume directory to block device(or global map?)

Reply:
I think that can work.  It seems like the SetUp is becoming more of a "local attach" method.  For attachable plugins like gce pd, I think SetUp will be a noop since it is already attached at a global path.

Reply:
>>  I think SetUp will be a noop since it is already attached at a global path.
yes, I haven't pushed my latest code, but setUp and tearDown now only do returning nil if plugin is attachable plugin.